### PR TITLE
feat: Radio Groups & Checkboxes

### PR DESCRIPTION
**Pull request by NeloBlivion**
## Summary
Implements radio groups, checkbox groups, and checkboxes in modals
image
Alpha server only right now

## Information

- [ ] This PR fixes an issue.
- [x] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed).
- [ ] This PR is **not** a code change (e.g. documentation, README, typehinting,
      examples, ...).

## Checklist

- [x] I have searched the open pull requests for duplicates.
- [x] If code changes were made then they have been tested.
  - [x] I have updated the documentation to reflect the changes.
- [ ] If `type: ignore` comments were used, a comment is also left explaining why.
- [x] I have updated the changelog to include these changes.

### CODE REVIEW

File: discord/types/components.py [hunk 1]
```diff
@@ -174,6 +174,42 @@ class FileUploadComponent(BaseComponent):
     required: NotRequired[bool]
 
 
+class RadioGroupOption(TypedDict):
+    value: str,
+    label: str,
```
Comment (Lulalaby) [inline discord/types/components.py]:
```suggestion
    value: str
    label: str
```

File: discord/types/components.py [hunk 2]
```diff
@@ -174,6 +174,42 @@ class FileUploadComponent(BaseComponent):
...
+class RadioGroupComponent(BaseComponent):
+    type: Literal[21]
+    custom_id: str
+    required: NotRequired[bool]
+    options: list[RadioGroupOption]
+
+
+class CheckboxGroupOption(TypedDict):
+    value: str,
+    label: str,
```
Comment (Lulalaby) [inline discord/types/components.py]:
```suggestion
    value: str
    label: str
```

**Review by Lulalaby — COMMENTED** [review]
SMH blindly copying

File: discord/types/components.py [hunk 3]
```diff
@@ -174,6 +174,43 @@ class FileUploadComponent(BaseComponent):
     required: NotRequired[bool]
 
 
+class RadioGroupOption(TypedDict):
+    value: str,
```
Comment (Lulalaby) [inline discord/types/components.py]:
```suggestion
    value: str
```

File: discord/types/components.py [hunk 4]
```diff
@@ -174,6 +174,43 @@ class FileUploadComponent(BaseComponent):
     required: NotRequired[bool]
 
 
+class RadioGroupOption(TypedDict):
+    value: str
+    value: str
```
Comment (Lulalaby) [inline discord/types/components.py]:
```suggestion
    value: str
```

File: CHANGELOG.md [hunk 5]
```diff
@@ -14,6 +14,8 @@ These changes are available on the `master` branch, but have not yet been releas
 
 - Added `.extension` attribute to emojis to get their file extension.
   ([#3055](https://github.com/Pycord-Development/pycord/pull/3055))
+- Added `RadioGroup`, `CheckboxGroup`, & `Checkbox` for modals.
```
Comment (JustaSqu1d) [inline CHANGELOG.md]:
```suggestion
- Added `RadioGroup`, `CheckboxGroup`, and `Checkbox` for modals.
```

File: discord/components.py [hunk 6]
```diff
@@ -1419,6 +1431,360 @@ def to_dict(self) -> FileUploadComponentPayload:
...
+    custom_id: Optional[:class:`str`]
+        The custom ID of the checkbox group that gets received during an interaction.
+    options: List[:class:`CheckboxGroupOption`]
+        A list of options that can be selected in this group.
+    min_values: Optional[:class:`int`]
+        The minimum number of options that must be selected.
+    max_values: Optional[:class:`int`]
+        The maximum number of options that can be selected.
+    required: Optional[:class:`bool`]
+        Whether the checkbox group requires a selection or not. Defaults to `True`.
```
Comment (JustaSqu1d) [inline discord/components.py]:
```suggestion
        Whether the checkbox group requires a selection or not. Defaults to ``True``.
```

File: discord/components.py [hunk 7]
```diff
@@ -1419,6 +1431,360 @@ def to_dict(self) -> FileUploadComponentPayload:
...
+    .. versionadded:: 2.7.1
+
+    Attributes
+    ----------
+    custom_id: Optional[:class:`str`]
+        The custom ID of the radio group that gets received during an interaction.
+    options: List[:class:`RadioGroupOption`]
+        A list of options that can be selected in this group.
+    required: Optional[:class:`bool`]
+        Whether the radio group requires a selection or not. Defaults to `True`.
```
Comment (JustaSqu1d) [inline discord/components.py]:
```suggestion
        Whether the radio group requires a selection or not. Defaults to ``True``.
```

**Review by plun1331 — DISMISSED** [review]
yeah that probably looks okay or something and i might have read all the code im not sure

goodnight

**Comment by Paillat-dev** [comment]
U mean I going to sleep now I woke upe 30 mins ago lol

File: discord/components.py [hunk 8]
```diff
@@ -1419,6 +1431,359 @@ def to_dict(self) -> FileUploadComponentPayload:
...
+    )
+
+    __repr_info__: ClassVar[tuple[str, ...]] = __slots__
+    versions: tuple[int, ...] = (2,)
+
+    def __init__(self, data: RadioGroupComponentPayload):
+        self.type = ComponentType.radio_group
+        self.id: int | None = data.get("id")
+        self.custom_id = data.get("custom_id")
+        self.options: list[RadioGroupOption] = [
```
Comment (Paillat-dev) [inline discord/components.py:1472]: This is not really relevant to now specifically, but while going through this I wondered if there's a reason why we don't use list(map()) instead of a comprehension

File: discord/ui/checkbox.py [hunk 9]
```diff
@@ -0,0 +1,115 @@
+from __future__ import annotations
```
Comment (Paillat-dev) [inline discord/ui/checkbox.py:25]: Copyrights

File: discord/ui/radio_group.py [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+    from ..types.components import RadioGroupComponent as RadioGroupComponentPayload
+
+
+class RadioGroup(ModalItem):
+    """Represents a UI Radio Group component.
+
+    .. versionadded:: 2.7.1
+
+    Parameters
+    ----------
```
Comment (Paillat-dev) [inline discord/ui/radio_group.py]:
```suggestion
    Attributes
    ----------
```

**Comment by NeloBlivion** [comment]
todo:
- update with `generate_underlying`

File: discord/components.py (hunk 8 above)
Comment (Paillat-dev) [reply discord/components.py:1472]: The answer btw is that list(map()) is ~10% slower

File: discord/ui/checkbox.py (hunk 9 above)
Comment (Paillat-dev) [reply discord/ui/checkbox.py:25]: Bump

**Comment by Paillat-dev** [comment]
<details>
<summary>Test Code</summary>

```py
import logging
import os
from typing import override

from dotenv import load_dotenv

import discord

load_dotenv()
logging.basicConfig(level=logging.INFO)

bot = discord.Bot(intents=discord.Intents.all())


class WowModal(discord.ui.DesignerModal):
    def __init__(self):
        super().__init__(timeout=None, title="PYCORD IS AMAZING")
        self.add_item(
            discord.ui.Label(
                label="Enable",
                description="Enable or disable the bot",
                item=discord.ui.Checkbox(default=False),
            )
        )

    @override
    async def callback(self, interaction: discord.Interaction):
        await interaction.respond("Button clicked!")


@bot.slash_command(name="test_modalsv5")
async def test_modal(ctx: discord.ApplicationContext):
    await ctx.send_modal(WowModal())


bot.run(os.getenv("TOKEN_2"))
```


</details> 

Fyi I encounter this:

```
Traceback (most recent call last):
  File "/home/paillat/Documents/pycord/pycord/discord/commands/core.py", line 138, in wrapped
    ret = await coro(arg)
          ^^^^^^^^^^^^^^^
  File "/home/paillat/Documents/pycord/pycord/discord/commands/core.py", line 1108, in _invoke
    await self.callback(ctx, **kwargs)
  File "/home/paillat/Documents/pycord/pycord/test_modals_v5.py", line 29, in test_modal
    await ctx.send_modal(WowModal())
                         ~~~~~~~~^^
  File "/home/paillat/Documents/pycord/pycord/test_modals_v5.py", line 18, in __init__
    self.add_item(discord.ui.Label(label="Enable", description="Enable or disable the bot", item=discord.ui.Checkbox(default=False)))
                  ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/paillat/Documents/pycord/pycord/discord/ui/label.py", line 124, in __init__
    self.set_item(item)
    ~~~~~~~~~~~~~^^^^^^
  File "/home/paillat/Documents/pycord/pycord/discord/ui/label.py", line 177, in set_item
    self._set_component_from_item(item)
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^
  File "/home/paillat/Documents/pycord/pycord/discord/ui/label.py", line 133, in _set_component_from_item
    self.underlying.component = item._generate_underlying()
                                ~~~~~~~~~~~~~~~~~~~~~~~~~^^
TypeError: Item._generate_underlying() missing 1 required positional argument: 'cls'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/paillat/Documents/pycord/pycord/discord/bot.py", line 1154, in invoke_application_command
    await ctx.command.invoke(ctx)
  File "/home/paillat/Documents/pycord/pycord/discord/commands/core.py", line 435, in invoke
    await injected(ctx)
  File "/home/paillat/Documents/pycord/pycord/discord/commands/core.py", line 146, in wrapped
    raise ApplicationCommandInvokeError(exc) from exc
discord.errors.ApplicationCommandInvokeError: Application Command raised an exception: TypeError: Item._generate_underlying() missing 1 required positional argument: 'cls'
```
```

**Comment by NeloBlivion** [comment]
probably done?

File: discord/components.py [hunk 11]
```diff
@@ -1419,6 +1431,359 @@ def to_dict(self) -> FileUploadComponentPayload:
...
+    def to_dict(self) -> CheckboxComponentPayload:
+        payload = {
+            "type": 23,
+            "custom_id": self.custom_id,
+        }
+        if self.id is not None:
+            payload["id"] = self.id
+
+        if self.default is not None:
+            payload["default"] = self.default
```
Comment (Soheab) [inline discord/components.py:1784]: Can't be None, should be in the initial payload.

File: discord/ui/checkbox.py [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+            custom_id=custom_id,
+            default=default,
+            id=id,
+        )
+
+    def __repr__(self) -> str:
+        attrs = " ".join(
+            f"{key}={getattr(self, key)!r}" for key in self.__item_repr_attributes__
+        )
+        return f"<{self.__class__.__name__} {attrs}>"
```
Comment (Soheab) [inline discord/ui/checkbox.py]: Why does it redefine this? It's handled by `Item`.

File: discord/components.py [hunk 13]
```diff
@@ -1419,6 +1431,359 @@ def to_dict(self) -> FileUploadComponentPayload:
...
+        This class is not useable by end-users; see :class:`discord.ui.RadioGroup` instead.
+
+    .. versionadded:: 2.7.1
+
+    Attributes
+    ----------
+    custom_id: Optional[:class:`str`]
+        The custom ID of the radio group that gets received during an interaction.
+    options: List[:class:`RadioGroupOption`]
+        A list of options that can be selected in this group.
```
Comment (Soheab) [inline discord/components.py]: Could mention that this can only contain min 2 and max 10 items.

File: discord/ui/checkbox_group.py [hunk 14]
```diff
@@ -0,0 +1,282 @@
...
+            max_values=max_values,
+            required=required,
+            id=id,
+        )
+
+    def __repr__(self) -> str:
+        attrs = " ".join(
+            f"{key}={getattr(self, key)!r}" for key in self.__item_repr_attributes__
+        )
+        return f"<{self.__class__.__name__} {attrs}>"
```
Comment (Soheab) [inline discord/ui/checkbox_group.py]: Same as checkbox

File: discord/ui/radio_group.py [hunk 15]
```diff
@@ -0,0 +1,234 @@
...
+            options=options or [],
+            required=required,
+            id=id,
+        )
+
+    def __repr__(self) -> str:
+        attrs = " ".join(
+            f"{key}={getattr(self, key)!r}" for key in self.__item_repr_attributes__
+        )
+        return f"<{self.__class__.__name__} {attrs}>"
```
Comment (Soheab) [inline discord/ui/radio_group.py]: Same as checkbox

File: discord/ui/radio_group.py [hunk 16]
```diff
@@ -0,0 +1,234 @@
...
+    def options(self) -> list[RadioGroupOption]:
+        """A list of options that can be selected in this group."""
+        return self.underlying.options
+
+    @options.setter
+    def options(self, value: list[RadioGroupOption]):
+        if not isinstance(value, list):
+            raise TypeError("options must be a list of RadioGroupOption")
+        if len(value) > 10:
+            raise ValueError("you may only provide up to 10 options.")
```
Comment (Soheab) [inline discord/ui/radio_group.py]: Maybe also check the minimum (2)?

**Comment by Soheab** [comment]
All version strings need changing to 2.8 too.

File: discord/ui/radio_group.py [hunk 17]
```diff
@@ -0,0 +1,234 @@
...
+        id: int | None = None,
+    ):
+        super().__init__()
+        if custom_id is not None and not isinstance(custom_id, str):
+            raise TypeError(
+                f"expected custom_id to be str, not {custom_id.__class__.__name__}"
+            )
+        if not isinstance(required, bool):
+            raise TypeError(f"required must be bool not {required.__class__.__name__}")  # type: ignore
+        custom_id = os.urandom(16).hex() if custom_id is None else custom_id
```
Comment (Paillat-dev) [inline discord/ui/radio_group.py:84]: Could be neat to have validation for there to be max(?) one option with default set to True

**Comment by Paillat-dev** [comment]
For the record, PR tested and everything seems to work fine other than obviously the comments

File: discord/ui/file_upload.py [hunk 18]
```diff
@@ -57,15 +81,15 @@ def __init__(
...
+            raise ValueError("max_values must be between 1 and 10")
         if custom_id is not None and not isinstance(custom_id, str):
             raise TypeError(
                 f"expected custom_id to be str, not {custom_id.__class__.__name__}"
             )
         if not isinstance(required, bool):
             raise TypeError(f"required must be bool not {required.__class__.__name__}")  # type: ignore
         custom_id = os.urandom(16).hex() if custom_id is None else custom_id
-        self._attachments: list[Attachment] | None = None
+        self._attachments: list[Attachment] = []
```
Comment (plun1331) [inline discord/ui/file_upload.py]: This was supposed to be `None` to differentiate between no interaction being recieved

Comment (NeloBlivion) [reply discord/ui/file_upload.py]: hmmm this is awkward, it seems the [select default values pr](https://github.com/Pycord-Development/pycord/commit/3b8b79f2040caa1bd4130fd4b01cde688e496261#diff-74a820e22350488813b4e6e66abfc721fd3bbbf0db28a27431410c44c3daa1ebR670) changed to to an empty list in selects and so I went off of that (even though the docs weren't updated); i'll revert both changes instead for consistency.

File: discord/components.py (hunk 11 above)
Comment (NeloBlivion) [reply discord/components.py:1784]: it's probably an edge case, but `default` is considered optional in the send payload and so it's reflected as such here

**Comment by Paillat-dev** [comment]
@Lumabots when you have time you can test this if you want

File: discord/ui/radio_group.py [hunk 19]
```diff
@@ -0,0 +1,238 @@
...
+    @property
+    def options(self) -> list[RadioGroupOption]:
+        """A list of options that can be selected in this group."""
+        return self.underlying.options
+
+    @options.setter
+    def options(self, value: list[RadioGroupOption]):
+        if not isinstance(value, list):
+            raise TypeError("options must be a list of RadioGroupOption")
+        if not (2 >= len(value) >= 10) and len(value) != 0:
```
Comment (CT-34589) [inline discord/ui/radio_group.py]: needs swapping to this:
`if not (2 <= len(value) <= 10) and len(value) != 0:`
as it currently returns true for any non-empty list

File: discord/ui/checkbox.py [hunk 20]
```diff
@@ -0,0 +1,139 @@
...
+
+    @property
+    def default(self) -> bool:
+        """Whether this checkbox is selected by default or not. Defaults to ``False``"""
+        return self.underlying.default
+
+    @default.setter
+    def default(self, value: bool):
+        if not isinstance(value, bool):
+            raise TypeError(f"default must be bool, not {value.__class__.__name__}")  # type: ignore
```
Comment (plun1331) [inline discord/ui/checkbox.py]: what's this ignored for?

File: discord/ui/modal.py [hunk 21]
```diff
@@ -158,7 +157,7 @@ def custom_id(self, value: str):
             raise TypeError(
                 f"expected custom_id to be str, not {value.__class__.__name__}"
             )
-        if len(value) > 100:
+        if value and len(value) > 100:
```
Comment (plun1331) [inline discord/ui/modal.py]: not really needed?

File: discord/ui/radio_group.py [hunk 22]
```diff
@@ -0,0 +1,238 @@
...
+    @property
+    def custom_id(self) -> str:
+        """The custom id that gets received during an interaction."""
+        return self.underlying.custom_id
+
+    @custom_id.setter
+    def custom_id(self, value: str):
+        if not isinstance(value, str):
+            raise TypeError(
+                f"custom_id must be None or str not {value.__class__.__name__}"
```
Comment (plun1331) [inline discord/ui/radio_group.py]:
```suggestion
                f"custom_id must be str not {value.__class__.__name__}"
```

File: discord/ui/select.py [hunk 23]
```diff
@@ -377,7 +377,7 @@ def custom_id(self) -> str:
     def custom_id(self, value: str):
         if not isinstance(value, str):
             raise TypeError("custom_id must be None or str")
-        if len(value) > 100:
+        if value and len(value) > 100:
```
Comment (plun1331) [inline discord/ui/select.py]: still not really needed

File: discord/ui/radio_group.py [hunk 24]
```diff
@@ -0,0 +1,238 @@
...
+            )
+        if value and len(value) > 100:
```
Comment (plun1331) [inline discord/ui/radio_group.py]: still not needed

File: discord/ui/modal.py (hunk 21 above)
Comment (NeloBlivion) [reply discord/ui/modal.py]: i see, i mixed it up because you can set the custom_id of `Button` to `None`, but this doesn't apply to any other item; reverted.

File: discord/ui/checkbox.py (hunk 20 above)
Comment (NeloBlivion) [reply discord/ui/checkbox.py]: not a clue, seems to be a holdover from other item implementations, removed
